framework/src/eventloop/eventloop_event.c

```diff
@@ -0,0 +1,358 @@
+			eventloop_unregister_event_cb(event);
+			return;
+		}
+	}
+	if (data->cb_info.event_data != NULL) {
+		EL_FREE(data->cb_info.event_data);
+	}
+	data->cb_info.event_data = NULL;
+	data->cb_info.event_data_size = 0;
+	data->event_in = -1;
+
+	return;
+}
+
+static int eventloop_send_event_sig(int type, void *event_data, int data_size)
+{
+	event_group_t *event_group;
+	event_node_t *ptr;
+
+	if (type < 0 || type >= EL_EVENT_MAX || data_size < 0) {
+		eldbg("Invalid Parameter\n");
+		return EVENTLOOP_INVALID_PARAM;
+	}
+
+	event_group = get_event_group(type);
+	if (event_group != NULL) {
+		ptr = (event_node_t *)sq_peek(&event_group->event_list);
+		while (ptr != NULL && ptr->event != NULL) {
+			event_data_t *cb_data = (event_data_t *)ptr->event->data;
+			cb_data->event_in = type;
+			if (data_size == 0) {
+				cb_data->cb_info.event_data = NULL;
+			} else {
+				cb_data->cb_info.event_data = EL_ALLOC(data_size);
+				if (cb_data->cb_info.event_data == NULL) {
+					eldbg("Failed to allocate callback info\n");
+					return EVENTLOOP_OUT_OF_MEMORY;
+				}
+				memcpy(cb_data->cb_info.event_data, event_data, data_size);
+			}
+			cb_data->cb_info.event_data_size = data_size;
+			/* Send signal to task which registered event */
+			int ret = kill(cb_data->pid, SIGEL_EVENT);
+			if (ret < 0) {
+				eldbg("kill failed %d \n", errno);
+				if (cb_data->cb_info.event_data != NULL) {
+					EL_FREE(cb_data->cb_info.event_data);
+					cb_data->cb_info.event_data = NULL;
+				}
+			}
+			ptr = (event_node_t *)sq_next(ptr);
+		}
+	}
+
+	return OK;
+}
+
+el_event_t *eventloop_add_event_handler(int type, event_callback func, void *data, int data_size)
+{
+	int ret;
+	el_loop_t *loop;
+	el_event_t *handle;
+	event_data_t *event_cb;
+
+	if (type < 0 || type >= EL_EVENT_MAX || func == NULL || data_size < 0) {
+		eldbg("Invalid Parameter\n");
+		return NULL;
+	}
+
+	loop = get_app_loop();
+	if (loop == NULL) {
+		eldbg("Failed to get loop\n");
+		return NULL;
+	}
+
+	handle = (el_event_t *)EL_ALLOC(sizeof(el_event_t));
+	if (handle == NULL) {
+		eldbg("Failed to allocate event\n");
+		return NULL;
+	}
+
+	event_cb = (event_data_t *)EL_ALLOC(sizeof(event_data_t));
+	if (event_cb == NULL) {
+		eldbg("Failed to allocate callback\n");
+		EL_FREE(handle);
+		return NULL;
+	}
+
+	event_cb->type = type;
+	event_cb->pid = getpid();
+	event_cb->event_in = -1;
+	event_cb->cb_info.func = func;
+	if (data_size == 0) {
+		event_cb->cb_info.cb_data = NULL;
+	} else {
+		event_cb->cb_info.cb_data = EL_ALLOC(data_size);
+		if (event_cb->cb_info.cb_data == NULL) {
+			eldbg("Failed to allocate callback info\n");
+			goto errout;
+		}
+		memcpy(event_cb->cb_info.cb_data, data, data_size);
+	}
+	event_cb->cb_info.cb_data_size = data_size;
+	event_cb->cb_info.event_data = NULL;
+	event_cb->cb_info.event_data_size = 0;
+	handle->data = (void *)event_cb;
+
+	ret = uv_signal_init(loop, handle);
+	if (ret != 0) {
+		eldbg("Failed to initialize event\n");
+		goto errout;

```

Comment:
I think `EL_FREE(event_cb->cb_info.cb_data)` is needed at here and other below `goto errout` cases

Reply:
Already the logic you said are in errout.